Components documentation improvements

HiDeoo opened the pull request:
#### Description

- Closes #1938 and #2010

This PR is a follow-up of #1938 to improve the components documentation and move away from the [single page](https://starlight.astro.build/guides/components/) currently used.

This pull request is a draft as it only includes a global component guide and the new references for the `<Tabs>` and `<Badge>` components. The idea is to first get feedback on the new format, discuss remaining questions and then continue with the rest of the components once the format is agreed upon.

The `<Badge>` page was chosen as it's relatively simple and small while the `<Tabs>` page was chosen as it's a more complex one which involves 2 components.

To give a brief overview of some of the changes per page:

##### Global Components Guide

- This is pretty much the [introduction](https://starlight.astro.build/guides/components/#using-a-component) of the current Components guide.
- It also includes a new section for `ComponentProps` to address #2010.

##### Component References

The global structure is as follows:

- Brief description of the component
- Preview of the component in a basic example with no code example
- Import statement
- A usage section with various use cases
  - Each usage includes a code example and the associated preview
- A props section with a list of all accepted props
  - A link to the component implementation (can be useful to see where/how extra props are spread for example)
  - This follow the pattern used in other references

Some other notes:

- Links exists between props mentioned in the usage section and the props section ~~and vice versa~~ (decided that it was not useful after discussing this).
- In the props section, the `required` tag is not included in the title (this allows for switching languages without losing position compared to the other reference pages).
- The component used to display code examples and/or previews is pretty much unstyled at this point of the PR:
  - The idea of the component is to make the examples more obvious with some kind of frame around it compared to the current examples.
  - We would need to decide if we want some controls on it or keep it as simple as possible:
    - I think one important thing to consider is that Starlight is not a component library/design system
    - Most examples researched include a "Show Code" / "Open in StackBlitz" button which I don't think are relevant for Starlight as we want to show both the code and the preview at the same time.
    - One of the controls I think could be relevant would be a toggle to switch the preview between light and dark mode although this one could be tricky to implement with Starlight's current code.
  - Do we want to keep the duplicated content in the markup (for the code and preview) like we currently have or do we want to have a more advanced setup, maybe relying on remark, to only have to specify the code once?
    - Not sure this is worth it considering the number of components and the complexity it could add

#### Remaining tasks

- [x] Once the format is agreed upon, convert the rest of the components to the new format
- [x] Use a custom UI string for the "Preview" word
  - [x] Wait for #1923 to be merged
- [x] Remove the existing components guide
- [x] Add a redirect for the old guide (`/guides/components/`)
- [x] Figure out where the new references should live in the sidebar (I used a top-level group for now to make it easier to find)
- [ ] Make sure to merge with the Lunaria tracker directive from [this comment](https://github.com/withastro/starlight/pull/2121#issuecomment-2353169456)

astrobot-houston commented:
<!-- lunaria-action-comment -->

## Lunaria Status Overview


🌕 **This pull request will trigger status changes.**

<details>
<summary>Learn more</summary>

By default, every PR changing files present in the [Lunaria configuration's `files` property](https://lunaria.dev/reference/configuration/#files-required) will be considered and trigger status changes accordingly.

You can change this by adding one of the keywords present in the [`ignoreKeywords` property in your Lunaria configuration file](https://lunaria.dev/reference/configuration/#ignorekeywords) in the PR's title (ignoring all files) or by [including a tracker directive](https://lunaria.dev/guides/tracking/#tracker-directives) in the merged commit's description. 
</details>

### Tracked Files




| Locale | File                                                                                                                                                                                    | Note                            |
| ------ | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------- |
| en     | [components/badges.mdx](https://github.com/withastro/starlight/blob/c9dacee9c25cccc53ac5a9844bcd9aed4734262c/docs%2Fsrc%2Fcontent%2Fdocs%2Fcomponents%2Fbadges.mdx)                     | Source added, will be tracked.  |
| en     | [components/tabs.mdx](https://github.com/withastro/starlight/blob/c9dacee9c25cccc53ac5a9844bcd9aed4734262c/docs%2Fsrc%2Fcontent%2Fdocs%2Fcomponents%2Ftabs.mdx)                         | Source added, will be tracked.  |
| en     | [components/using-components.mdx](https://github.com/withastro/starlight/blob/c9dacee9c25cccc53ac5a9844bcd9aed4734262c/docs%2Fsrc%2Fcontent%2Fdocs%2Fcomponents%2Fusing-components.mdx) | Source added, will be tracked.  |

<details>
<summary>Warnings reference</summary>

| Icon | Description                                                                                                                                       |
| ---- | ------------------------------------------------------------------------------------------------------------------------------------------------- |
| 🔄️  | The source for this localization has been updated since the creation of this pull request, make sure all changes in the source have been applied. |

</details>

HiDeoo commented:
Updated the pull request, here is a summary of the changes:

- Updated the existing page structure based on feedback gathered during Talking & Doc'ing
- Added new component pages for all existing Starlight components
  - One potential area I'm not really sure about is the `<Card>`, `<LinkCard>`, and `<CardGrid>` components, which are now 3 different pages. I think it makes sense as `<CardGrid>` can be used with both and we probably don't want to duplicate the information in both the `<Card>` and `<LinkCard>` pages. But I'm open to feedback on this.
- Added a new reference page for icons (based on Sarah's idea)
  - As icons are not always used with `<Icon>` but can be used a prop in other components, or even frontmatter values, the icons list is now a separate reference page.
  - Searching using the browser's search feature is more reliable as the page mostly consists of icon names.
  - As the idea is that icons are not only tied to the `<Icon>` component, clicking on an icon will now copy the icon name to the clipboard and not the entire `<Icon>` component with the icon name. (this has probably been one of my most annoying issues with the current implementation, I always want the name, never the icon. It makes testing different icons a pain as you have to replace the entire `<Icon>` component with your new clipboard value, and if your `<Icon>` use any extra prop, e.g. `size`, you cannot easily paste at all)
  - I tried implementing in this page an icon categorization based on Unicons category (as we discussed [here](https://github.com/withastro/starlight/pull/1713#pullrequestreview-1983073155)), but reverted the change as I don't think it's a good solution at the moment. As we only have a small selection of icons from the Unicons library, we can end up with some categories with only one icon. I think we should wait until we embed the full Unicons library or use `astro-icon` to use a categorization based on categories.

HiDeoo commented:
Updated the PR with the changes of #2087 & #1784.

HiDeoo commented:
In anticipation of #2249, I updated the PR with Markdoc support:

- Added a script to generate Markdoc grammar from the [one](https://github.com/markdoc/language-server) in their own LSP server repo and configured the documentation to use it (`markdoc` language identifier)
- Updated the ["Using components"](https://deploy-preview-2121--astro-starlight.netlify.app/components/using-components/) page to mention Markdoc (it's the one I am the most unsure about)
- Updated all component pages to also include Markdoc examples (basic `<Tabs>` at the moment, maybe need to refine the UI a bit)
  - The only example not having a Markdoc equivalent is the ["Use imported code"](https://deploy-preview-2121--astro-starlight.netlify.app/components/code/#use-imported-code) one as I don't think it's doable.
  - Formatting wise, I used the "Format" button on their [playground](https://markdoc.dev/sandbox?mode=preview) for all examples.

<img width="743" alt="image" src="https://github.com/user-attachments/assets/5b4d3fe3-c4d7-4851-9176-693ba39ee81c">

lorenzolewis commented on docs/src/content/docs/components/using-components.mdx:
````diff
@@ -0,0 +1,79 @@
+title: Welcome to my docs
+---
+
+import SomeComponent from '../../components/SomeComponent.astro';
+import AnotherComponent from '../../components/AnotherComponent.astro';
+
+<SomeComponent prop="something" />
+
+<AnotherComponent>
+	Components can also contain **nested content**.
+</AnotherComponent>
+```
````
It would be nice to give some more concrete examples here (such as `<WrapperComponent />` for the last component). Thanks to https://www.rainsberger.ca/blog/50-43-whats-in-a-name/#_top for bringing this to mind.

lorenzolewis commented on docs/src/content/docs/components/using-components.mdx:
```diff
@@ -0,0 +1,79 @@
+
+See the components group in the sidebar for a list of available components and how to use them.
+
+## Compatibility with Starlight’s styles
```
```suggestion
## Compatibility with Starlight styles
```

Not sure if the apostrophe is strictly needed here

delucis commented on docs/src/content/docs/components/file-tree.mdx:
```diff
@@ -0,0 +1,237 @@
+Display a file tree with file icons and collapsible sub-directories using the `<FileTree>` component.
+
+Specify the structure of your files and directories with an [unordered Markdown list](https://www.markdownguide.org/basic-syntax/#unordered-lists) inside `<FileTree>`.
+Create a sub-directory using a nested list or add a `/` to the end of a list item to render it as a directory without specific content.
```
We no longer show this `directory/` syntax in this simplified example, but it might make sense to bring it back given we don’t show it elsewhere?

delucis commented on docs/src/content/docs/components/file-tree.mdx:
```diff
@@ -0,0 +1,237 @@
+- src
+  - components
+    - Header.astro
+    - Title.astro
```
Could be as simple as adding back this from the opening example:

```suggestion
    - Title.astro
  - pages/
```

delucis commented on docs/src/content/docs/components/file-tree.mdx:
````diff
@@ -0,0 +1,237 @@
+- src
+  - components
+    - Header.astro
+    - Title.astro
+
+</FileTree>
+```
+
+<Fragment slot="markdoc">
+
+```markdoc
+# src/content/docs/example.mdoc
+
+{% filetree %}
+- astro.config.mjs
+- package.json
+- src
+  - components
+    - Header.astro
+    - Title.astro
+{% /filetree %}
+```
+
+</Fragment>
+
+<FileTree slot="preview">
+
+- astro.config.mjs
+- package.json
+- src
+  - components
+    - Header.astro
+    - Title.astro
````
Could be as simple as adding back the `pages/` directory from the opening example:

````suggestion
    - Title.astro
  - pages/

</FileTree>
```

<Fragment slot="markdoc">

```markdoc
# src/content/docs/example.mdoc

{% filetree %}
- astro.config.mjs
- package.json
- src
  - components
    - Header.astro
    - Title.astro
  - pages/
{% /filetree %}
```

</Fragment>

<FileTree slot="preview">

- astro.config.mjs
- package.json
- src
  - components
    - Header.astro
    - Title.astro
  - pages/
````

delucis commented on docs/src/content/docs/components/file-tree.mdx:
```diff
@@ -0,0 +1,237 @@
+### Add comments
+
+Add a comment to a file or directory by adding more text after the name.
+The comment can be any text, including Markdown formatting.
```
I might focus more on the formatting and underline _inline_ options — you can’t use add a paragraph etc.

```suggestion
Inline Markdown formatting such as bold and italics is supported in comments.
```

delucis commented on docs/src/content/docs/components/file-tree.mdx:
```diff
@@ -0,0 +1,237 @@
+
+### Add placeholders
+
+Add placeholder files and directories by using either `...` or `…` as the name.
```
This feels like it might benefit from a small explanation for _why_ you might want to show a placeholder:

```suggestion
Add placeholder files and directories by using either `...` or `…` as the name.
This can be useful to indicate to a reader that a folder is expected to contain more items without specifying them all explicitly.
```

delucis commented on docs/src/content/docs/components/icons.mdx:
```diff
@@ -0,0 +1,132 @@
+
+import { Icon } from '@astrojs/starlight/components';
+
+To display icons from a set of [common icons](/reference/icons/#all-icons), use the `<Icon>` component.
```
Couple of nits here:
- you can’t use “a set” — there’s only one supported set currently, so we should use a definite article (or like in this case specify “our” icons
- I think rather than “common”, which in English often means “widespread” or “frequent”, maybe something like “built-in” is a better way to describe these (mirrors the component language too “built-in components” / “built-in icons”)

```suggestion
To display icons from Starlight’s [built-in icon set](/reference/icons/#all-icons), use the `<Icon>` component.
```

delucis commented on docs/src/content/docs/components/icons.mdx:
```diff
@@ -0,0 +1,132 @@
+## Usage
+
+Display an icon using the `<Icon>` component.
+An icon requires a [`name`](#name) set to [one of Starlight’s built-in icons](/reference/icons/#all-icons) and can optionally include a [`label`](#label) to provide context for screen readers.
```
Ha, you got it perfect here already! 😄

delucis commented on docs/src/content/docs/components/icons.mdx:
```diff
@@ -0,0 +1,132 @@
+
+**type:** `string`
+
+An optional label to provide context for screen readers.
```
Might be worth documenting the full accessibility behaviour here while we’re at it:

```suggestion
An optional label to provide context for assistive technologies, such as screen readers.

When `label` is not set, the icon will be completely hidden from assistive technologies.
In this case, make sure the context is still understandable without the icon.
For example, a link containing only the icon **must** include the `label` attribute in order to be accessible, but if a link contains text and the icon is purely decorative, omitting the `label` may make sense.
```

delucis commented on docs/src/content/docs/components/icons.mdx:
```diff
@@ -0,0 +1,132 @@
+
+**type:** `string`
+
+Defines custom CSS classes to add to the icon.
```
Micro-nit, but the other entries just say “the size…”, “the color…”, so we can probably skip the “defines” here too:

```suggestion
Custom CSS classes to add to the icon.
```

delucis commented on docs/src/content/docs/reference/icons.mdx:
```diff
@@ -0,0 +1,19 @@
+description: An overview of all the icons available in Starlight.
+---
+
+Starlight provides a set of common icons that you can display in your content using the `<Icon>` component.
```
Same comment as in the `<Icon>` page re: “common”:

```suggestion
Starlight provides a set of built-in icons that you can display in your content using the `<Icon>` component.
```

delucis commented on docs/src/content/docs/components/link-buttons.mdx:
```diff
@@ -0,0 +1,152 @@
+
+## Usage
+
+Display a link button, useful for directing users to the most relevant or actionable content and often used on landing pages, using the `<LinkButton>` component.
```
While it means we repeat ourselves slightly between the opening and this section, I think this still flows better so it’s worth doing:
```suggestion
Use the `<LinkButton>` component to display a visually distinct call-to-action link. A link button is useful for directing users to the most relevant or actionable content and is often used on landing pages
```

delucis commented on docs/src/content/docs/components/link-buttons.mdx:
```diff
@@ -0,0 +1,152 @@
+
+**Implementation:** [`LinkButton.astro`](https://github.com/withastro/starlight/blob/main/packages/starlight/user-components/LinkButton.astro)
+
+The `<LinkButton>` component accepts the following props and also any other `<a>` attributes:
```
I think it would be nice to link here like the `LinkCard` page does:

```suggestion
The `<LinkButton>` component accepts the following props and also any [other `<a>` attributes](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a):
```

delucis commented on docs/src/content/docs/components/link-buttons.mdx:
```diff
@@ -0,0 +1,152 @@
+
+### `href`
+
+**required**
```
Missing line break here!

```suggestion
**required**<br>
```

delucis commented on docs/src/content/docs/components/steps.mdx:
```diff
@@ -0,0 +1,124 @@
+
+import { Tabs, TabItem, Steps } from '@astrojs/starlight/components';
+
+To style a numbered lists of tasks to create step-by-step guides, use the `<Steps>` component.
```
Typo:
```suggestion
To style a numbered list of tasks to create step-by-step guides, use the `<Steps>` component.
```

delucis commented on docs/src/content/docs/components/tabs.mdx:
```diff
@@ -0,0 +1,232 @@
+
+import { Tabs, TabItem } from '@astrojs/starlight/components';
+
+To create a tabbed interface useful to organize content by grouping similar information together, use the `<Tabs>` and `<TabItem>` components.
```
Can we sneak in my “tabs are for equivalents” line here? 😁 

```suggestion
To create a tabbed interface use the `<Tabs>` and `<TabItem>` components.
Tabs are useful for displaying equivalent information where a user only needs to see one of several options.
```

delucis reviewed (commented):
Such incredible work — reviewing it in detail really drove home the attention to detail. Really appreciate it! (Still can’t quite believe you generated a Markdoc grammar _just_ for this 😁)

Thank you @HiDeoo!

yanthomasdev commented on docs/src/content/docs/components/asides.mdx:
```diff
@@ -0,0 +1,160 @@
+**type:** `'note' | 'tip' | 'caution' | 'danger'`  
+**default:** `'note'`
+
+The type of aside to display: `note` (the default), `tip`, `caution` or `danger`.
```
If we go the simpler route, we're missing a comma.
```suggestion
The type of aside to display: `note` (the default), `tip`, `caution`, or `danger`.
```

yanthomasdev commented on docs/src/content/docs/components/asides.mdx:
```diff
@@ -0,0 +1,160 @@
+**type:** `'note' | 'tip' | 'caution' | 'danger'`  
+**default:** `'note'`
+
+The type of aside to display: `note` (the default), `tip`, `caution` or `danger`.
```
I guess we should add a full stop at those bullet items? (since it's a full sentence)

yanthomasdev commented on docs/src/content/docs/components/badges.mdx:
```diff
@@ -0,0 +1,156 @@
+---
+title: Badges
+description: Learn how to use badges in Starlight.
```
The other description gives a small indication of what you'd use the component for, something like this maybe?

```suggestion
description: Learn how to use badges in Starlight to display additional pieces of information.
```

yanthomasdev commented on docs/src/content/docs/components/card-grids.mdx:
```diff
@@ -0,0 +1,183 @@
+---
+title: Card Grids
+description: Learn how to wrap multiple cards in a grid.
```
Since other descriptions have "in Starlight" I would add that to all descriptions to keep it consistent.
```suggestion
description: Learn how to wrap multiple cards in a grid in Starlight.
```

yanthomasdev commented on docs/src/content/docs/components/cards.mdx:
```diff
@@ -0,0 +1,116 @@
+---
+title: Cards
+description: Learn how to display content in a box with the Card component.
```
Maybe we could rewrite the description to more closely resemble the others?
```suggestion
description: Learn how to cards in Starlight to display content in a box.
```

yanthomasdev commented on docs/src/content/docs/components/code.mdx:
```diff
@@ -0,0 +1,106 @@
+---
+title: Code
+description: Learn how to display syntax highlighted code using the Code component.
```
I feel like the main reason you'd use this is when you can't use the Markdown code block, so worth mentioning it in the description?
```suggestion
description: Learn how to display syntax highlighted code in Starlight without Markdown code blocks.
```

yanthomasdev commented on docs/src/content/docs/components/file-tree.mdx:
```diff
@@ -0,0 +1,237 @@
+---
+title: File Tree
+description: Learn how to display the structure of a directory with file icons and collapsible sub-directories.
```
```suggestion
description: Learn how to display the structure of a directory with file icons and collapsible sub-directories in Starlight.
```

yanthomasdev commented on docs/src/content/docs/components/link-cards.mdx:
```diff
@@ -0,0 +1,133 @@
+---
+title: Link Cards
+description: Learn how to display links prominently with the LinkCard component.
```
```suggestion
description: Learn how to display links prominently as cards in Starlight.
```

yanthomasdev commented on docs/src/content/docs/components/steps.mdx:
```diff
@@ -0,0 +1,124 @@
+---
+title: Steps
+description: Learn how to style numbered lists of tasks to create step-by-step guides.
```
```suggestion
description: Learn how to style numbered lists of tasks to create step-by-step guides in Starlight.
```

yanthomasdev commented on docs/src/content/docs/components/tabs.mdx:
```diff
@@ -0,0 +1,232 @@
+---
+title: Tabs
+description: Learn how to create tabbed interfaces in Starlight.
```
```suggestion
description: Learn how to create tabbed interfaces in Starlight to display equivalent information closely.
```

delucis replied:
Not sure about the "closely" here. Maybe something like "...to group equivalent information"?

yanthomasdev commented on docs/src/content/docs/components/tabs.mdx:
```diff
@@ -0,0 +1,232 @@
+Keep multiple tab groups synchronized by adding the [`syncKey`](#synckey) attribute.
+
+All `<Tabs>` on a page with the same `syncKey` value will display the same active label.
+This allows your reader to choose once (e.g. their operating system or package manager), and see their choice persisted across page navigations.
```
```suggestion
This allows your reader to choose once (e.g. their operating system or package manager), and see their choice persist across page navigations.
```

delucis replied:
Heh, funny one. I actually think both can be correct in this context and have a slight preference for "persisted" while struggling to articulate why 😄 

It's really subtle but "see ... persist" is like an active "I am watching this thing persist all the time" vs "see ... persisted" being a slightly more passive "the thing is persisted by the website, I'll notice when I notice". Very much a _feeling_ of the language though, rather than either being grammatically incorrect.

yanthomasdev commented on docs/src/content/docs/components/using-components.mdx:
```diff
@@ -0,0 +1,79 @@
+
+## Compatibility with Starlight’s styles
+
+Starlight applies default styling to your Markdown content, for example adding margin between elements.
```
```suggestion
Starlight applies default styling to your Markdown content, for example, adding margin between elements.
```

yanthomasdev reviewed (commented):
Great work, I have a few suggestions to consider 🙌

HiDeoo commented:
Thanks everyone for the great reviews, the branch should now be updated with all the changes.

I also removed the old component guides which required a few extra changes:

- A lot of tests were using that slug, so I updated them to use `guides/project-structure` instead (as we still want a page in the "Guides" group)
- There was a few references to the page to update.

An interesting point with the removal of the old component guide is that we have a few languages that don't have this guide, so we would no longer get fallback content for it, but they have a few links to it which will cause the link validation to fail:

```
[ERROR] ✗ Found 2 invalid links in 2 files.
▶ pt-pt/getting-started/
  └─ /pt-pt/guides/components/ - invalid link
▶ uk/getting-started/
  └─ /uk/guides/components/ - invalid link
```

I guess we would want to update these 2 `getting-started` pages to link to the new localized version of the `components/using-components` page (which will be fallback content), but we would want to Lunaria ignore these 2 changes? Am I correct?

HiDeoo commented:
The following files were updated to fix invalid links and should be Lunaria ignored:

- `docs/src/content/docs/pt-pt/getting-started.mdx`
- `docs/src/content/docs/uk/getting-started.mdx`

**When merged, this PR should include the following Lunaria tracker directive:**

```
@lunaria-ignore:src/content/docs/pt-pt/getting-started.mdx;src/content/docs/uk/getting-started.mdx
```

sarah11918 commented on docs/src/content/docs/components/using-components.mdx:
```diff
@@ -0,0 +1,94 @@
+Starlight provides built-in components for common documentation use cases.
+These components are available from the `@astrojs/starlight/components` package in MDX files and from the [Starlight Markdoc preset](/guides/authoring-content/#markdoc) in Markdoc files.
+
+See the components group in the sidebar for a list of available components and how to use them.
```
```suggestion
See the components group in the sidebar for a list of available components and how to use them.
```
"the components group" in the sidebar feels off to me. If someone is on this page, they'll already be in this section in the sidebar. And, I know "group" is the way we define the sidebar sections, but I don't think someone reading/navigating a page thinks of "group" as identifying where they are. Heading? Section maybe?

I don't know the history of how this line came to be, so I'm hesitant to make a suggestion that might have been overruled previously. Just calling it out! I at first for a moment didn't realize you were pointing me to look in the sidebar, I thought there was some "component group" as a thing somewhere I needed to find. So I think the terminology just feels a little non-standard for someone *reading* this docs site, as opposed to building one.

delucis replied:
No big history! The vague inspiration was this in the getting started guide:

> See the guides in the sidebar for help adding content and customizing your Starlight site.

But no “group” mention there either.

HiDeoo replied:
Would saying "See the sidebar for a list of available components and how to use them." be too vague or confusing?

sarah11918 replied:
I think that's just fine!

HiDeoo replied:
Thanks a lot, just updated the wording 🌟

sarah11918 commented on docs/src/content/docs/components/card-grids.mdx:
````diff
@@ -0,0 +1,173 @@
+
+<Preview>
+
+```mdx {5,12}
````
```suggestion
```mdx {5,8}
```
I think this is the intention? In the preview, only one line is highlighted and I wondered why!

HiDeoo replied:
Great catch, thanks a lot, it was actually all examples in all formats having wrong lines/ranges since I removed the filenames.

delucis reviewed (approved):
Awesome work on this @HiDeoo! From the nice new preview pane design to the careful work structuring these pages in a consistent way that is flexible across all the different components we have, this is lightyears ahead of where we were. Stellar work 💖
